doc/oggzdump.1
...\" $Header: /var/lib/cvsd/annodex/liboggz/doc/oggzdump.1,v 1.1 2004/01/04 05:28:25 zen Exp $
...\"
...\"	transcript compatibility for postscript use.
...\"
...\"	synopsis:  .P! <file.ps>
...\"
.de P!
.fl
\!!1 setgray
.fl
\\&.\"
.fl
\!!0 setgray
.fl			\" force out current output buffer
\!!save /psv exch def currentpoint translate 0 0 moveto
\!!/showpage{}def
.fl			\" prolog
.sy sed -e 's/^/!/' \\$1\" bring in postscript file
\!!psv restore
.
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
'\" t 
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n  
.TH "oggzdump" "1" 
.SH "NAME" 
oggzdump \(em Hexdump packets of an Ogg file, or revert an Ogg file from 
such a hexdump 
 
.SH "SYNOPSIS" 
.PP 
\fBoggzdump\fR [-b  | --binary ]  [-x  | --hexadecimal ]  [-n  | --new ]  [-o \fBfilename\fR  | --output \fBfilename\fR ]  [-s \fBserialno\fR  | --serialno \fBserialno\fR ]  [-O  | --hide-offset ]  [-S  | --hide-serialno ]  [-G  | --hide-granulepos ]  [-P  | --hide-packetno ]  [-h  | --help ] filename  
.PP 
\fBoggzdump\fR [-r  | --revert ]  [-o \fBfilename\fR  | --output \fBfilename\fR ] filename  
.SH "Description" 
.PP 
\fBoggzdump\fR generates a hexadecimal or binary dump 
of packets in an Ogg file, or creates an Ogg file from a hexadecimal 
oggzdump file. 
 
.SH "Options" 
.PP 
\fBoggzdump\fR accepts the following options: 
 
.SS "Dump format options" 
.IP "-b, --binary" 10 
Generate a binary dump of each packet 
 
.IP "-x, --hexadeciaml" 10 
Generate a hexadecimal dump of each packet 
 
.SS "Filtering options" 
.IP "-n, --new" 10 
Only dump the first packet of each new logical 
bitstream found; for correctly formed Ogg files, this dumps 
only the beginning-of-stream packets. 
 
.IP "-s \fBserialno\fR, --serialno \fBserialno\fR" 10 
Dump only the logical bitstream with specified 
\fBserialno\fR. 
 
.IP "-O, --hide-offset" 10 
Hide the byte offset field of the preamble for 
each packet dumped. This is particularly useful in 
conjunction with 
\fBoggzdiff\fP\fB(1)\fP when comparing Ogg bitstreams that differ 
only in page structure. 
 
.IP "-S, --hide-serialno" 10 
Hide the serialno field of the preamble for 
each packet dumped. This is particularly useful in 
conjunction with 
\fBoggzdiff\fP\fB(1)\fP when comparing Ogg bitstreams that differ 
only in serialno assignment. 
 
.IP "-G, --hide-granulepos" 10 
Hide the granulepos field of the preamble for 
each packet dumped. 
 
.IP "-P, --hide-packetno" 10 
Hide the packetno field of the preamble for 
each packet dumped. 
 
.SS "Mode options" 
.IP "-r, --revert" 10 
Revert an oggzdump. Generates an Ogg bitstream as prescribed 
in the input oggzdump. 
 
.IP "" 10 
The serialno, granulepos and packetno fields of the packet 
preambles in the oggzdump, as well as any bos or eos markers 
if present, are used to structure the packets into logical 
bitstreams. Hence it is advised that the original input 
oggzdump not be generated with any hiding filters. 
 
.IP "" 10 
Only the hexadecimal data columns of the input oggzdump are 
parsed when generating Ogg packet data; the printable character 
representation is ignored. 
 
.SS "Miscellaneous options" 
.IP "-o \fBfilename\fR, --output \fBfilename\fR" 10 
Write output to the specified 
\fBfilename\fR instead of printing it to 
standard output. 
 
.IP "-h, --help" 10 
Display usage information and exit. 
.SH "AUTHOR" 
.PP 
Conrad Parker        November 17, 2003;      
.SH "COPYRIGHT" 
.PP 
Copyright \(co 2003 CSIRO Australia 
 
.SH "SEE ALSO" 
.PP 
\fBoggzdiff\fP\fB(1)\fP, 
\fBxxd\fP\fB(1)\fP, 
\fBhexdump\fP\fB(1)\fP      
...\" created by instant / docbook-to-man, Wed 19 Nov 2003, 11:46 


ChangeLog
Thu Mar 11 11:00:00 EST 2004 Silvia Pfeiffer <silvia.pfeiffer@csiro.au>
	* Version 0.8.2
	* fixed up the Makefiles to not include the CVS subdirectories
	
Sun Mar 07 16:00:00 EST 2004 Silvia Pfeiffer <silvia.pfeiffer@csiro.au>
	* Version 0.8.1
	* includes Windows port with Makefile & VC6 workspace & .NET solution
	
Thu Oct 16 21:55:07 EST 2003 Conrad Parker <conrad@annodex.net>

	* split liboggz out from libannodex sources, started ChangeLog


src/tools/oggzdump.c
/*
   Copyright (C) 2003 Commonwealth Scientific and Industrial Research
   Organisation (CSIRO) Australia

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:

   - Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.

   - Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

   - Neither the name of CSIRO Australia nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
   PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE ORGANISATION OR
   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#ifdef HAVE_INTTYPES_H
#  include <inttypes.h>
#else
#  define PRId64 "I64d"
#endif

#include <getopt.h>
#include <errno.h>

#include <oggz/oggz.h>
#include "oggz_tools.h"

/*#define DEBUG*/

#undef MIN
#define MIN(a,b) (((a)<(b))?(a):(b))

typedef struct {
  OggzTable * content_types_table;
  OggzReadPacket read_packet;
} ODData;

static char * progname;
static FILE * outfile = NULL;
static int dump_bits = 0;
static int dump_char = 1;
static int truth = 1;

static int hide_offset = 0;
static int hide_serialno = 0;
static int hide_granulepos = 0;
static int hide_packetno = 0;

static void
usage (char * progname)
{
  printf ("Usage: %s [options] filename\n", progname);
  printf ("Hexdump packets of an Ogg file, or revert an Ogg file from such a hexdump\n");
  printf ("\nDump format options\n");
  printf ("  -b, --binary           Generate a binary dump of each packet\n");
  printf ("  -x, --hexadecimal      Generate a hexadecimal dump of each packet\n");
  printf ("\nFiltering options\n");
  printf ("  -n, --new              Only dump the first packet of each logical bitstream\n");
  printf ("  -c content-type, --content-type content-type\n");
  printf ("                         Dump only the logical bitstreams for a specified\n");
  printf ("                         content-type. The following codecs are currently\n");
  printf ("                         detected: theora, vorbis, speex, cmml\n");
  printf ("  -s serialno, --serialno serialno\n");
  printf ("                         Dump only the logical bitstream with specified serialno\n");
  printf ("  -O, --hide-offset      Hide the byte offset of each packet\n");
  printf ("  -S, --hide-serialno    Hide the serialno field of each packet\n");
  printf ("  -G, --hide-granulepos  Hide the granulepos field of each packet\n");
  printf ("  -P, --hide-packetno    Hide the packetno field of each packet\n");
  printf ("\nMode options\n");
  printf ("  -r, --revert           Revert an oggzdump. Generates an Ogg bitstream\n");
  printf ("                         as prescribed in the input oggzdump\n");
  printf ("\nMiscellaneous options\n");
  printf ("  -o filename, --output filename\n");
  printf ("                         Specify output filename\n");
  printf ("  -h, --help             Display this help and exit\n");
  printf ("  -v, --version          Output version information and exit\n");
  printf ("\n");
  printf ("Please report bugs to <ogg-dev@xiph.org>\n");
}

static ODData *
oddata_new ()
{
  ODData * oddata = malloc (sizeof (ODData));
  memset (oddata, 0, sizeof (ODData));

  oddata->content_types_table = oggz_table_new ();

  return oddata;
}

static void
oddata_delete (ODData * oddata)
{
  oggz_table_delete (oddata->content_types_table);

  free (oddata);
}

static void
dump_char_line (unsigned char * buf, long n)
{
  int i;

  fprintf (outfile, "  ");

  for (i = 0; i < n; i++) {
    if (isgraph(buf[i])) fprintf (outfile, "%c", buf[i]);
    else if (isspace(buf[i])) fprintf (outfile, " ");
    else fprintf (outfile, ".");
  }
}

static void
hex_dump (unsigned char * buf, long n)
{
  int i;
  long remaining = n, count = 0;
  long rowlen;

  while (remaining > 0) {
    rowlen = MIN (remaining, 16);

    if (n > 0xffffff)
      fprintf (outfile, "%08lx:", count);
    else if (n > 0xffff)
      fprintf (outfile, "  %06lx:", count);
    else
      fprintf (outfile, "    %04lx:", count);

    for (i = 0; i < rowlen; i++) {
      if (!(i%2)) fprintf (outfile, " ");
      fprintf (outfile, "%02x", buf[i]);
    }

    for (; i < 16; i++) {
      if (!(i%2)) fprintf (outfile, " ");
      fprintf (outfile, "  ");
    }

    if (dump_char)
      dump_char_line (buf, rowlen);

    fprintf(outfile, "\n");

    remaining -= rowlen;
    buf += rowlen;
    count += rowlen;
  }
}

static void
bin_dump (unsigned char * buf, long n)
{
  int i, j;
  long remaining = n, count = 0;
  long rowlen;

  while (remaining > 0) {
    rowlen = MIN (remaining, 6);

    if (n > 0xffffff)
      fprintf (outfile, "%08lx:", count);
    else if (n > 0xffff)
      fprintf (outfile, "  %06lx:", count);
    else
      fprintf (outfile, "    %04lx:", count);

    for (i = 0; i < rowlen; i++) {
      fprintf (outfile, " ");
#ifdef WORDS_BIGENDIAN
      for (j = 0; j < 8; j++)
#else
      for (j = 7; j >= 0; j--)
#endif
	fprintf (outfile, "%c", (buf[i]&(1<<j)) ? '1' : '0');
    }

    for (; i < 6; i++) {
      if (!(i%2)) fprintf (outfile, " ");
      fprintf (outfile, "         ");
    }

    if (dump_char)
      dump_char_line (buf, rowlen);

    printf("\n");

    remaining -= rowlen;
    buf += rowlen;
    count += rowlen;
  }
}

/* FIXME: on Mac OS X, off_t is 64-bits.  Obviously we want a nicer
 * way to do it than this, but a quick fix is a good fix */
#ifdef __APPLE__
#  define PRI_off_t "q"
#else
#  define PRI_off_t "l"
#endif

static int
read_packet (OGGZ * oggz, ogg_packet * op, long serialno, void * user_data)
{
  ogg_int64_t units;
  double time_offset;

  units = oggz_tell_units (oggz);
  if (hide_offset) {
    fprintf (outfile, "oOo");
  } else if (units == -1) {
    fprintf (outfile, "%08" PRI_off_t "x", oggz_tell (oggz));
  } else {
    time_offset = (double)units / 1000.0;
    ot_fprint_time (outfile, time_offset);
  }

  fprintf (outfile, ": serialno %010ld, "
	   "granulepos %" PRId64 ", packetno %" PRId64,
	   hide_serialno ? -1 : serialno,
	   hide_granulepos ? -1 : op->granulepos,
	   hide_packetno ? -1 : op->packetno);

  if (op->b_o_s) {
    fprintf (outfile, " *** bos");
  }

  if (op->e_o_s) {
    fprintf (outfile, " *** eos");
  }

  fprintf (outfile, ": ");
  ot_fprint_bytes (outfile, op->bytes);
  fputc ('\n', outfile);

  if (dump_bits) {
    bin_dump (op->packet, op->bytes);
  } else {
    hex_dump (op->packet, op->bytes);
  }

  fprintf (outfile, "\n");

  return 0;
}

static int
filter_page (OGGZ * oggz, const ogg_page * og, long serialno, void * user_data)
{
  ODData * oddata = (ODData *) user_data;
  const char * ident;
  int i, n;

  if (ogg_page_bos ((ogg_page *)og)) {
    ident = ot_page_identify (og, NULL);
    n = oggz_table_size (oddata->content_types_table);
    for (i = 0; i < n; i++) {
      char * c = oggz_table_nth (oddata->content_types_table, i, NULL);
      if (strcasecmp (c, ident) == 0) {
	oggz_set_read_callback (oggz, serialno, oddata->read_packet, NULL);
      }
    }
  }

  return 0;
}

static int
ignore_packet (OGGZ * oggz, ogg_packet * op, long serialno, void * user_data)
{
  return -1;
}

static int
read_new_packet (OGGZ * oggz, ogg_packet * op, long serialno, void * user_data)
{
  oggz_set_read_callback (oggz, serialno, ignore_packet, NULL);
  read_packet (oggz, op, serialno, user_data);

  return -1;
}

static void
revert_file (char * infilename)
{
  OGGZ * oggz;
  FILE * infile;
  char line[80];
  int hh, mm, ss;
  unsigned int offset;
  long current_serialno = -1, serialno;
  ogg_int64_t granulepos, packetno;
  int bos = 0, eos = 0;

  int line_offset = 0, consumed = 0;

  unsigned char * packet = NULL;
  long max_bytes = 0;

  unsigned char buf[1024];
  ogg_packet op;
  int flush = 1;
  long n;
  char c;

  if (strcmp (infilename, "-") == 0) {
    infile = stdin;
  } else {
    infile = fopen (infilename, "rb");
  }

  oggz = oggz_new (OGGZ_WRITE|OGGZ_NONSTRICT);

  while (fgets (line, 80, infile)) {
    line_offset = 0;

    /* Skip time offsets, OR ensure line_offset is 0 */
    if (sscanf (line, "%2d:%2d:%2d.%n", &hh, &mm, &ss, &line_offset) < 3)
      line_offset = 0;

    if (sscanf (&line[line_offset], "%x: serialno %ld, granulepos %lld, packetno %lld%n",
		&offset, &serialno, &granulepos, &packetno,
		&line_offset) >= 4) {

      /* flush any existing packets */
      if (current_serialno != -1) {
	int ret;

#ifdef DEBUG
	printf ("feeding packet (%010ld) %ld bytes\n",
		current_serialno, op.bytes);
#endif
	if ((ret = oggz_write_feed (oggz, &op, current_serialno, flush, NULL)) != 0) {
	  fprintf (stderr, "%s: oggz_write_feed error %d\n", progname, ret);
	}

	while ((n = oggz_write_output (oggz, buf, 1024)) > 0) {
	  fwrite (buf, 1, n, outfile);
	}
      }

      /* Start new packet */
      bos = 0; eos = 0;
      if (sscanf (&line[line_offset], " *** %[b]%[o]%[s]%n", &c, &c, &c,
		  &consumed) >= 3) {
	bos = 1;
	line_offset += consumed;
      }
      if (sscanf (&line[line_offset], " *** %[e]%[o]%[s]%n", &c, &c, &c,
		  &consumed) >= 3) {
	eos = 1;
	line_offset += consumed;
      }

      current_serialno = serialno;

      op.packet = packet;
      op.bytes = 0;
      op.b_o_s = bos;
      op.e_o_s = eos;
      op.granulepos = granulepos;
      op.packetno = packetno;

    } else {
      int nread = 0;
      unsigned int val = 0;
      unsigned int offset;

      if (current_serialno != -1 &&
	  sscanf (line, "%x:%n", &offset, &line_offset) >= 1) {
	while (nread < 16 &&
	       (sscanf (&line[line_offset], "%2x%n", &val, &consumed) > 0)) {
	  op.bytes++;
	  if (op.bytes > max_bytes) {
	    unsigned char * new_packet;
	    size_t new_size;

	    if (max_bytes == 0) {
	      new_size = 128;
	    } else {
	      new_size = max_bytes * 2;
	    }

	    new_packet =
	      (unsigned char *) realloc ((void *)packet, new_size);
	    if (new_packet == NULL) {
	      fprintf (stderr,
		       "%s: error allocating memory for packet data\n",
		       progname);
	      exit (1);
	    } else {
	      max_bytes = (long)new_size;
	      packet = new_packet;
	      op.packet = packet;
	    }
	  }

	  packet[op.bytes-1] = (unsigned char) val;

	  line_offset += consumed;
	  nread++;
	}
      }
    }
  }

  fclose (infile);
}

int
main (int argc, char ** argv)
{
  int show_version = 0;
  int show_help = 0;

  ODData * oddata;
  OGGZ * oggz;
  char * infilename = NULL, * outfilename = NULL;
  int revert = 0;
  OggzTable * table = NULL;
  long serialno;
  int i, size;
  long n;

  int filter_serialnos = 0;
  int filter_content_types = 0;

  ot_init ();

  progname = argv[0];

  if (argc < 2) {
    usage (progname);
    return (1);
  }

  oddata = oddata_new ();

  oddata->read_packet = read_packet;

  table = oggz_table_new();

  while (1) {
    char * optstring = "hvbxnro:s:c:OSGP";

#ifdef HAVE_GETOPT_LONG
    static struct option long_options[] = {
      {"help", no_argument, 0, 'h'},
      {"version", no_argument, 0, 'v'},
      {"binary", no_argument, 0, 'b'},
      {"hexadecimal", no_argument, 0, 'x'},
      {"new", no_argument, 0, 'n'},
      {"revert", no_argument, 0, 'r'},
      {"output", required_argument, 0, 'o'},
      {"serialno", required_argument, 0, 's'},
      {"content-type", required_argument, 0, 'c'},
      {"hide-offset", no_argument, 0, 'O'},
      {"hide-serialno", no_argument, 0, 'S'},
      {"hide-granulepos", no_argument, 0, 'G'},
      {"hide-packetno", no_argument, 0, 'P'},
      {0,0,0,0}
    };

    i = getopt_long(argc, argv, optstring, long_options, NULL);
#else
    i = getopt (argc, argv, optstring);
#endif
    if (i == -1) break;
    if (i == ':') {
      usage (progname);
      goto exit_err;
    }

    switch (i) {
    case 'h': /* help */
      show_help = 1;
      break;
    case 'v': /* version */
      show_version = 1;
      break;
    case 'b': /* binary */
      dump_bits = 1;
      break;
    case 'n': /* new */
      oddata->read_packet = read_new_packet;
      break;
    case 'o': /* output */
      outfilename = optarg;
      break;
    case 'r': /* revert */
      revert = 1;
      break;
    case 's': /* serialno */
      filter_serialnos = 1;
      serialno = atol (optarg);
      oggz_table_insert (table, serialno, &truth);
      break;
    case 'c': /* content-type */
      filter_content_types = 1;
      n = (long)oggz_table_size (oddata->content_types_table);
      oggz_table_insert (oddata->content_types_table, (long)n, optarg);
      break;
    case 'O': /* hide offset */
      hide_offset = 1;
      break;
    case 'S': /* hide serialno */
      hide_serialno = 1;
      break;
    case 'G': /* hide granulepos */
      hide_granulepos = 1;
      break;
    case 'P': /* hide packetno */
      hide_packetno = 1;
      break;
    default:
      break;
    }
  }

  if (show_version) {
    printf ("%s version " VERSION "\n", progname);
  }

  if (show_help) {
    usage (progname);
  }

  if (show_version || show_help) {
    goto exit_ok;
  }

  if (optind >= argc) {
    usage (progname);
    goto exit_err;
  }

  infilename = argv[optind++];

  if (outfilename == NULL) {
    outfile = stdout;
  } else {
    outfile = fopen (outfilename, "wb");
    if (outfile == NULL) {
      fprintf (stderr, "%s: unable to open output file %s\n",
	       progname, outfilename);
      goto exit_err;
    }
  }

  if (revert) {
    if (dump_bits) {
      fprintf (stderr, "%s: Revert of binary dump not supported\n", progname);
      goto exit_err;
    }

    revert_file (infilename);
  } else {
    errno = 0;

    if (strcmp (infilename, "-") == 0) {
      oggz = oggz_open_stdio (stdin, OGGZ_READ|OGGZ_AUTO);
    } else {
      oggz = oggz_open (infilename, OGGZ_READ|OGGZ_AUTO);
    }

    if (oggz == NULL) {
      if (errno == 0) {
	fprintf (stderr, "%s: %s: OGGZ error opening input file\n",
		 progname, infilename);
      } else {
	fprintf (stderr, "%s: %s: %s\n",
		 progname, infilename, strerror (errno));
      }
      goto exit_err;
    }

    if (!filter_serialnos && !filter_content_types) {
      oggz_set_read_callback (oggz, -1, oddata->read_packet, NULL);
    } else {
      if (filter_serialnos) {
	size = oggz_table_size (table);
	for (i = 0; i < size; i++) {
	  oggz_table_nth (table, i, &serialno);
	  oggz_set_read_callback (oggz, serialno, oddata->read_packet, NULL);
	}
      }
      
      if (filter_content_types) {
	oggz_set_read_page (oggz, -1, filter_page, oddata);
      }
    }

    while ((n = oggz_read (oggz, 1024)) > 0);

    oggz_close (oggz);
  }

 exit_ok:
  oddata_delete (oddata);
  oggz_table_delete (table);
  exit (0);

 exit_err:
  oddata_delete (oddata);
  oggz_table_delete (table);
  exit (1);
}
